File: test_shared/src/main/java/com/clevertap/android/shared/test/BaseTestCase.java
package com.clevertap.android.shared.test;

import android.os.Build;
import android.os.Bundle;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.clevertap.android.sdk.BaseCTApiListener;
import com.clevertap.android.sdk.CleverTapAPI;
import com.clevertap.android.sdk.CleverTapInstanceConfig;

import org.junit.Assert;
import org.junit.Before;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.robolectric.annotation.Config;

import static com.clevertap.android.shared.test.Constant.ACC_ID;
import static com.clevertap.android.shared.test.Constant.ACC_TOKEN;
import static org.mockito.Mockito.when;

@Config(manifest = Config.NONE, sdk = {Build.VERSION_CODES.P},
        application = TestApplication.class
)
@RunWith(AndroidJUnit4.class)
public abstract class BaseTestCase {

    protected CleverTapAPI cleverTapAPI;
    protected TestApplication application;
    protected CleverTapInstanceConfig cleverTapInstanceConfig;
    protected BaseCTApiListener baseCTApiListener;

    public static void assertBundlesEquals(Bundle expected, Bundle actual) {
        assertBundlesEquals(null, expected, actual);
    }

    public static void assertBundlesEquals(String message, Bundle expected, Bundle actual) {
        if (!areEqual(expected, actual)) {
            Assert.fail(message + " <" + expected.toString() + "> is not equal to <" + actual.toString() + ">");
        }
    }

    public static boolean areEqual(Bundle expected, Bundle actual) {
        if (expected == null) {
            return actual == null;
        }

        if (expected.size() != actual.size()) {
            return false;
        }

        for (String key : expected.keySet()) {
            if (!actual.containsKey(key)) {
                return false;
            }

            Object expectedValue = expected.get(key);
            Object actualValue = actual.get(key);

            if (expectedValue == null) {
                if (actualValue != null) {
                    return false;
                }

                continue;
            }

            if (expectedValue instanceof Bundle && actualValue instanceof Bundle) {
                if (!areEqual((Bundle) expectedValue, (Bundle) actualValue)) {
                    return false;
                }

                continue;
            }

            if (!expectedValue.equals(actualValue)) {
                return false;
            }
        }

        return true;
    }

    @Before
    public void setUp() {
        application = TestApplication.getApplication();
        cleverTapAPI = Mockito.mock(CleverTapAPI.class);

        cleverTapInstanceConfig = CleverTapInstanceConfig.createInstance(application, ACC_ID, ACC_TOKEN);

        baseCTApiListener = Mockito.mock(BaseCTApiListener.class);
        when(baseCTApiListener.context()).thenReturn(application);
        when(baseCTApiListener.config()).thenReturn(cleverTapInstanceConfig);
    }

    public TestApplication getApplication() {
        return TestApplication.getApplication();
    }

}

File: test_shared/src/main/java/com/clevertap/android/shared/test/BaseTestCase.kt
package com.clevertap.android.shared.test

import android.content.Context
import android.os.Build.VERSION_CODES
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.clevertap.android.sdk.CleverTapAPI
import com.clevertap.android.sdk.CleverTapInstanceConfig
import io.mockk.clearAllMocks
import io.mockk.mockk
import org.json.JSONArray
import org.json.JSONObject
import org.junit.After
import org.junit.Before
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.Shadows
import org.robolectric.android.controller.ActivityController
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowApplication


/**
 * Naming Convention for Testing
 * 1. Classes : <Name of Class to be Test> + Test.kt
 *      e.g CTProductConfigControllerTest.kt for CTProductConfigController.java
 *
 * 2. Methods : test_<methodName>_<inputCondition>_<expectedBehavior>
 *     e.g test_constructor_whenFeatureFlagIsNotSave_InitShouldReturnTrue
 */
@Config(manifest = Config.NONE, sdk = [VERSION_CODES.P], application = TestApplication::class)
@RunWith(AndroidJUnit4::class)
abstract class BaseTestCase {

    protected lateinit var application: TestApplication
    protected lateinit var shadowApplication: ShadowApplication
    protected lateinit var cleverTapAPI: CleverTapAPI
    protected lateinit var cleverTapInstanceConfig: CleverTapInstanceConfig
    protected lateinit var activityController: ActivityController<TestActivity>
    protected lateinit var appCtx:Context

    @Before
    open fun setUp() {
        application = TestApplication.application
        shadowApplication = Shadows.shadowOf(application)
        cleverTapAPI = mockk<CleverTapAPI>(relaxed = true)
        cleverTapInstanceConfig = CleverTapInstanceConfig.createInstance(
            application,
            Constant.ACC_ID,
            Constant.ACC_TOKEN
        ).apply {
            // intentional
            customHandshakeDomain = null
        }
        activityController = Robolectric.buildActivity(TestActivity::class.java)
        appCtx = application.applicationContext

    }

    @After
    open fun cleanUp() {
        clearAllMocks()
    }


    fun getSampleJsonArrayOfJsonObjects(totalJsonObjects: Int = 0, start:Int=1, printGenArray:Boolean=false): JSONArray {

        val range = JSONArray()
        (start until start+totalJsonObjects).forEach {
            val obj = JSONObject()
            obj.put("key$it", it)
            range.put(obj)
        }
        return range.also { if(printGenArray)println("generated : $range") }
    }

    fun getSampleJsonArrayOfStrings(totalJsonObjects: Int = 0, start:Int=1, printGenArray:Boolean=false): JSONArray {

        val range = JSONArray()
        (start until start+totalJsonObjects).forEach {
            range.put("value$it")
        }
        return range.also { if(printGenArray)println("generated : $range") }
    }
}
